### PR TITLE
Avoid leak from user incorrectly holding to authCallback's callback.

### PR DESCRIPTION
**Pull request by tcard**
This is a corner case, but users may encounter it.

### CODE REVIEW

**Comment by QuintinWillison** [comment]
I'm having a look locally and may just end up pushing a couple of tweaks to save time.

File: Source/ARTAuth.m [hunk 1]
```diff
@@ -391,6 +391,26 @@ - (void)requestToken:(ARTTokenParams *)tokenParams withOptions:(ARTAuthOptions *
...
+            // to Rest/Realtime.
+            //
+            // So we're going to apply the Fundamental Theorem of Software Engineering here and wrap the
+            // callback in another reference. Then, we pass a callback that calls this reference,
+            // and not the original callback. Then, we return a task that, if cancelled, sets that
+            // intermediate reference to nil, so that we lose our reference to the original callback
+            // and release any references it may hold to Rest/Realtime. Even if the user holds on to
+            // the callback we give to them, they would be then holding to a now-nil reference to the
+            // original callback.
+            __weak void (^safeCallback)(id<ARTTokenDetailsCompatible> tokenDetailsCompat, NSError * error);
```
Comment (tcard) [inline Source/ARTAuth.m]: Why is this weak? Which reference cycle is this breaking?

File: Source/ARTTypes.m [hunk 2]
```diff
@@ -308,3 +308,74 @@ + (NSDate *)dateWithMillisecondsSince1970:(uint64_t)msecs {
...
+    }
+    self = [super init];
+    if (!self) {
+        return nil;
+    }
+    
+    _lock = [NSObject new];
+    _callback = callback;
+    
+    __weak ARTCancellableFromCallback * _cancellable = self;
```
Comment (tcard) [inline Source/ARTTypes.m]: Same, why is this weak?

File: Source/ARTAuth.m (hunk 1 above)
Comment (QuintinWillison) [reply Source/ARTAuth.m]: Just being defensive as I don't think the current scope has any business in 'owning' that callback. On other words, erring on the side of caution, in which case my question in return would be, why _should_ it be strong? It feels like a delegation pattern.

File: Source/ARTTypes.m (hunk 2 above)
Comment (QuintinWillison) [reply Source/ARTTypes.m]: This one is more important and I believe justified. We don't want `_wrapper` retaining a strong reference to `self`, given it's a strong ivar on `self` itself. Unless I missed something?

**Comment by QuintinWillison** [comment]
@tcard is there any chance my responses satisfied you and you're happy to resolve the two conversations above so I can get this merged?

**Comment by tcard** [comment]
@QuintinWillison Sorry, I got drowned in other things. LGTM

**Pull request merged**